### ISSUE TITLE
[DEV] Inaccurate use of bytecode maniuplations (ASM / Java Assist)

### ISSUE THREAD

**Issue by Elix-x** [ValkyrienWarfare/Valkyrien-Warfare-Revamped#46]
Here is the deal - while you were thinking of moving to Java Assist, i took a look at what you're actually patching. All i can say is that it is not always necessary. In many cases (i have not checked all of them), a forge event or other less severe type of call interception can be used. For example for `boolean EntityLivingBase#isOnLadder()` - a forge event can be used instead.

Am i saying that we can replace everything? I don't know. I have not looked at everything yet, but so far i have found other ways (than bytecode manipulations) of fixing it.

Can I do it? Yes. Will it take a lot of time? Yes, because i am already working on a literal ton of projects...

And if we will still have to use ASM / Java Assist? ASM would be easier to hook in as it is already done with forge. Hooking Java Assist is not a big problem. You'll just have to use ASM hooks and adapt them for Java Assist.

As for integrating Java Assist WITH obfuscation - i can't say anything. I have a lot more experience with ASM then with Java Assist, so i don't know how that can be worked out, other than by looking at its' docs.

TLDR: Hold on moving to Java Assist for now. There ARE things that DO NOT require any bytecode manipulation.

**Comment by Elix-x** [comment]
`Things like rendering ships in their respect block layer cannot be done, as there is no RenderBlockLayer event`
a. You can use other things than events (overriding `EntityRenderer` / `RenderGlobal` instance with your override in this case).
b. Submit event PR to forge.
`As for Java assist, it comes with some helpful methods for manipulating byte code, such as changing superclasses`
Because ASM is lower level then Java Assist, it can apply all changes Java Assist can (and possibly more).
But there is a very specific reason why MCF team choose low level ASM library rather than higher level one (Java Assist or even AspectJ) - to maintain as much mod compatibility with bytecode edits. And why conflicts can be caused is simple - if 2 mods edit the same method in the same class using Java Assist by rewriting it in pure java, the 2 mods become incompatible. Either the game will crash or one (or both) of mods won't work. By using low level API, you can modify the 1 and only line (or a few) you care about. In that case, the chance of conflicts is A LOT lower - the other mod has to modify exactly the same line&opcodes in order for one to occur.
And all the obfuscation. Even if you write modified method in pure java and don't care about all caused conflicts, you will have hard times (even harder than with ASM) obfuscating it in obfuscated env.

According to what you said, ASM patches are not written by you (and they are written quite poorly, for mc environment*). Heh. I no longer wonder why it crashes with Sponge.
AKA: You don't have to move to Java Assist to fix conflicts. Simply using ASM correctly is enough.

PS: MC Environment requires different methods of using ASM. Because it's special /:eye:\\.

**Comment by BigBastard** [comment]
Update: 
        There's still a lot of ASM invocation replaces being used by VW (perhaps too many), but this is down from when you originally posted it. Not everything can/should be handled by more complicated solutions than this in my opinion, but through modifying the Stack using Opcodes.DUP2_X2 and Opcodes.POP I've managed to setup a system that allows me to call my method upon the invocation of another, without replacing the original invocation. 

Anyways I'm curious of what you think of the current state of ASM in VW, as the sooner I can close this issue, the better I feel about Mod Compatibility overall.

**Comment by Elix-x** [comment]
Well, where possible ASM should be replaced with reflection/other hackery. I'll take a deeper look at what else can be replaced and how.

...And i guess how you're doing ASM too (because you saying that you are still intercepting method calls is a bit worrying)...

**Comment by Best108** [comment]
Old issue, closed.
